File: backend/typescript-node-express-hive/src/shared/web/error-handler.ts
import type { Request, Response, NextFunction } from "express";
import { ZodError } from "zod";
import { AppError, ErrorType } from "../errors/app-error.js";

export function errorHandler(err: any, req: Request, res: Response, next: NextFunction) {
    if (err instanceof AppError) {
        switch (err.type) {
            case ErrorType.NotFound:
                return respond(res, 404, err.message);
            case ErrorType.AlreadyExists:
                return respond(res, 422, err.message);
            case ErrorType.InvalidCredentials:
            case ErrorType.Unauthorized:
                return respond(res, 401, err.message);
            case ErrorType.Forbidden:
                return respond(res, 403, err.message);
            case ErrorType.Unprocessable:
                return respond(res, 422, err.message);
            case ErrorType.Internal:
                console.error(err.message);
                return respond(res, 500, err.message);
            default:
                return respond(res, 500, "An unexpected error occurred");
        }
    }

    if (err instanceof ZodError) {
        const messages = err.issues.map((e) => `${e.path.join(".")} ${e.message}`);
        return respondMultiple(res, 422, messages);
    }

    // Generic fallback for other errors
    console.error(err);
    const message = err instanceof Error ? err.message : "Internal Server Error";
    respond(res, 500, message);
}

export function respond(res: Response, code: number, message: string) {
    respondMultiple(res, code, [message]);
}

export function respondMultiple(res: Response, code: number, messages: string[]) {
    res.status(code).json({
        errors: {
            body: messages,
        },
    });
}

File: backend/python-fastapi-vertical/app/shared/web/error_handler.py
from fastapi import Request
from fastapi.responses import JSONResponse

from shared.errors import AppError, ErrorType


status_code_map = {
    ErrorType.NOT_FOUND: 404,
    ErrorType.ALREADY_EXISTS: 422,
    ErrorType.INVALID_CREDENTIALS: 422,
    ErrorType.UNAUTHORIZED: 401,
    ErrorType.FORBIDDEN: 403,
    ErrorType.UNPROCESSABLE: 422,
    ErrorType.INTERNAL: 500,
}


async def app_error_handler(request: Request, exc: AppError) -> JSONResponse:
    status_code = status_code_map.get(exc.error_type, 500)
    return JSONResponse(
        status_code=status_code,
        content={"errors": {"body": [exc.message]}},
    )


async def generic_error_handler(request: Request, exc: Exception) -> JSONResponse:
    return JSONResponse(
        status_code=500,
        content={"errors": {"body": [str(exc)]}},
    )
